### PR TITLE
Ember2 upgrade asscap

### PR DESCRIPTION
**Pull request by thescubageek**
* Adds Asset Uploader cap to Ember 2.0 Upgrade

PR: https://www.pivotaltracker.com/story/show/130773685
CMS: https://github.com/g5search/g5-content-management-system/pull/2079

### CODE REVIEW

File: app/components/s3-upload.js [hunk 1]
```diff
@@ -72,5 +78,106 @@ export default EmberUploader.FileField.extend({
   flashFor(promise) {
     return promise.then((successResp) => { this.successMessage(successResp); })
       .catch((failResp) => { this.dangerMessage(failResp); });
+  },
+
+  buildFilesArray(files){
+    this.set('files', files);
+    this.set('filesArray', this.convertFilesArray(files));
+    this.validateFileSizes();
+    return this.get('filesArray').length !== 0;
```
Comment (jcstringer) [inline app/components/s3-upload.js]: You can actually just call this.get('filesArray').length which is truthy if it is > 0 and falsey if it is 0.

File: app/components/s3-upload.js [hunk 2]
```diff
@@ -72,5 +78,106 @@ export default EmberUploader.FileField.extend({
...
+
+  validateFileSizes(){
+    this.tooBigWarningMessage(this.findImageFilesLargerThan(this.get('tooBigSize')));
+    this.bigWarningMessage(this.findImageFilesLargerThan(this.get('bigSize')));
+  },
+
+  findImageFilesLargerThan(size){
+    var results = [];
+    var filesArray = this.get('filesArray');
+    for (let i = 0; i < filesArray.length; i++){
```
Comment (jcstringer) [inline app/components/s3-upload.js]: Minor nitpick but you are recalculating length on every loop iteration, just set it once up front in a let and then reference it in your for loop.

File: app/components/s3-upload.js [hunk 3]
```diff
@@ -72,5 +78,106 @@ export default EmberUploader.FileField.extend({
...
+    }
+  },
+
+  warningMessage(warning, fileList, showConfirm=true){
+    var msg = `${warning}\n${this.fileListNames(fileList)}`;
+    return showConfirm ? `${msg}\nAre you sure you want to continue?` : msg;
+  },
+
+  removeFiles(fileList){
+    for (let i = 0; i < fileList.length; i++){
```
Comment (jcstringer) [inline app/components/s3-upload.js]: Ditto on calling length here.

File: app/components/s3-upload.js [hunk 4]
```diff
@@ -72,5 +78,106 @@ export default EmberUploader.FileField.extend({
...
+    }
+  },
+
+  removeAllFiles(){
+    this.set('filesArray', []);
+  },
+
+  fileListNames(list){
+    var names = "";
+    Ember.$.each(list, (idx, file)=>{
```
Comment (jcstringer) [inline app/components/s3-upload.js]: Ember has its own enumerable, you don't need to use the one in jquery. Not a big deal but checkout 
http://emberjs.com/api/classes/Ember.Enumerable.html#method_forEach

File: app/components/s3-upload.js [hunk 5]
```diff
@@ -72,5 +78,106 @@ export default EmberUploader.FileField.extend({
...
+      let size = this.formatSize(file.size);
+      names += `  ${file.name} (${size})\n`;
+    });
+    return names;
+  },
+
+  isImageFile(file){
+    var fileExt = file.name.split('.').pop().toLowerCase();
+    var imgExts = ['jpg','png','jpeg','gif'];
+    return Ember.$.inArray(fileExt, imgExts) !== -1;
```
Comment (jcstringer) [inline app/components/s3-upload.js]: http://emberjs.com/api/classes/Ember.Enumerable.html#method_includes

File: app/components/s3-upload.js [hunk 6]
```diff
@@ -72,5 +78,106 @@ export default EmberUploader.FileField.extend({
...
+  },
+
+  formatSize(size){
+    var byteUnits = ['kB', ' MB'];
+    for (var i = 0; i < byteUnits.length; i++){
```
Comment (jcstringer) [inline app/components/s3-upload.js]: Ditto

**Comment by jcstringer** [comment]
Doh no tests!
It can be annoying and hard to test this uploader stuff, but we can prob at least unit tests some if not all of the methods you wrote. Check these out for inspiration, you could prob just add to them. 

https://github.com/G5/orion-cms/tree/ember2-upgrade/tests/unit/components

File: app/components/s3-upload.js (hunk 2 above)
Comment (thescubageek) [inline app/components/s3-upload.js]: You know, I've _always_ done for loops with the `.length` assertion in the middle and never thought about the lookup overhead (which is minor but still there). Good call, I'll change my best practices to declare the length in a let variable

Comment (jcstringer) [inline app/components/s3-upload.js]: Ya me too, it's only in the last few years of doing a lot of Ember stuff that I've started trying to use best practices like that. Granted we don't have large collections of data in these kinds of loops so the end user would prob never know the difference in terms of performance, but I still try to use a local var for length as a best practice.

**Comment by thescubageek** [comment]
@jcstringer Specs should be finished, as should the tweaks you suggested in the review. If this looks good we should be fine to merge this in.

File: tests/unit/components/s3-upload-test.js [hunk 7]
```diff
@@ -81,3 +99,147 @@ test("didUploadHandler on the last file", function(assert){
...
+    assert.equal(msg, message, "show too big file size alert");
+  };
+  component.tooBigWarningMessage(tooBigFiles);
+  assert.deepEqual(component.get('filesArray'), [], "removes all files from list");
+});
+
+test("isImageFile detects image files", function(assert){
+  let component = this.subject({websiteId: 9});
+  let ret = component.isImageFile({name: "/path/to/some/img.jpg", size: 1048576});
+  assert.equal(ret, true, "identifies jpg files as images");
```
Comment (jcstringer) [inline tests/unit/components/s3-upload-test.js]: For boolean checks you can shorten these to 
```
assert.ok(ret, "identifies jpg files as images");
assert.ok(!ret, "identifies non-image files as not images");
```

**Comment by jcstringer** [comment]
This looks really good thanks for the test coverage.

**Comment by levibrown** [comment]
you can tie in the new alert styles and confirmations like this: https://github.com/G5/orion-cms/blob/ember2-upgrade/app/components/theme-config.js#L8
make sure to import sweet alert too

File: tests/unit/components/s3-upload-test.js (hunk 7 above)
Comment (thescubageek) [inline tests/unit/components/s3-upload-test.js]: Thanks good to know, I've updated the code accordingly

**Comment by thescubageek** [comment]
@levibrown @jcstringer Converting the alert/confirmation messages over to Sweet Alert proved to involve a lot more than just a simple function swap, as Sweet Alert's asynchronous way of handling confirmation/alert is _very_ different than the blocking `confirm`/`alert` on the window level. I have created another branch with these changes: https://github.com/G5/orion-cms/pull/124 ... still need to finish specs tomorrow as this is another big series of changes

**Comment by jcstringer** [comment]
🚢 I merged in your other branch to this one @thescubageek. If you would not mind, please fix the merge conflict in the css then go ahead and merge this in once you come back online later today.

**Pull request merged**